app/controllers/links_controller.rb

```diff
@@ -0,0 +1,43 @@
+class LinksController < ApplicationController
+  def edit
+    @local_authority = LocalAuthority.find_by(slug: params[:local_authority_slug])
+    @service = Service.find_by(slug: params[:service_slug])
+    @interaction = Interaction.find_by(slug: params[:interaction_slug])
+
+    @link = Link.get_link(params)
+    if @link.nil?
+      @link = Link.new
+      @link.url = 'n/a'
+    end
+  end
+
+  def update
+    link = Link.get_link(params)
+    link.url = params[:link][:url]
+    validate_and_save(link)
+  end
+
+  def create
+    link = Link.new
+    link.local_authority = LocalAuthority.find_by(slug: params[:local_authority_slug])
+    link.service = Service.find_by(slug: params[:service_slug])
+    link.interaction = Interaction.find_by(slug: params[:interaction_slug])
+    link.service_interaction = ServiceInteraction.find_by(service: link.service, interaction: link.interaction)
+    link.url = params[:link][:url]
+
+    validate_and_save(link)
+  end
+
+private
+
+  def validate_and_save(link)
+    if link.valid?
+      link.save!
+      flash[:success] = "Link has been saved."
+      redirect_to local_authority_service_interactions_path(service_slug: params[:service_slug])
+    else
+      flash[:danger] = "Please enter a valid link."
+      redirect_to action: "edit"
```

Comment:
Because we redirect here we lose a load of information about what the user has done.

1. We lose the url that they entered - so when we render the form again we can only show them the url from the db (or "n/a")
2. We lose the `link` instance that we called `.valid?` on which means we lose the error messages on there - when rails renders a form using `form_for` the form helpers will look at the object we passed to that to decide if the input elements should be given error classes to style them differently.

This is because a redirect causes a whole new http request/response to happen and the `link` instance and user entered params only exist for the request/response they were created in.  If instead of redirecting we asked to render the edit form again (`render :edit`) we'd keep the `link` instance so that rendering the form could use the validation errors and the url that the user had entered already.  Note that the form wants to use `@link` but our `create` action only uses `link`...